### DISCUSSION TITLE
How can I make a DiGraph for heterograph with different node types and edge types?

### DISCUSSION THREAD

**Discussion by Jeriousman** [networkx/networkx#5360] [Q&A] [answered]
How can I make a DiGraph for heterograph with different node types and edge types?

I would like node to have different types like node1 type1, node type 2, node type3 etc..
Just like the node type, I want edges to have different types.
And, like PyG or DGL, how can I make node types to have independent indices? example would be as below:
```
Node user: [0, 1, 2, 3, 4, 5, 6]
Node movie: [0, 1, 2, 3, 4, 5, 6, 7, 8]
Node actor: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
Not like node 0 ~ 6 for user, 7 ~ 15 for movie, 8 ~ 18 actor. Can someone please guide me with some example code? Thank you.

**Comment by dschult** [comment, marked as answer]
Nodes in NetworkX can be any object you like. And they can have arbitrary attributes attached to them.
Which way you want to encode the nodes and attributes probably depends more on what you want to do with them than what is easiest to store.  You could make the nodes be strings that encode the  type "movie-3" for example. But perhaps better to use unique node numbers across all types and use an attribute like `G.add_nodes_from([0, 1, 2, 3, 4], type="movie")`. It's very flexible.

**Reply by Jeriousman** [reply]
Thank you for the reply. But G.add_nodes_from([0, 1, 2, 3, 4], type="movie"), G.add_nodes_from([0, 1, 2, 3, 4], type="user") will make overwrite the nodes. But then there is no way to make node indices go from 0 onwards for each unique node types?
I am not so familiar with networkX I was rather familiar with PyG and DGL so.. wish you can inform me in a more detailed way. Much appreciated!

**Reply by dschult** [reply]
Yes, nodes have to be unique...   My first suggestion is one way to make them unique (here with more detail):
```
G.add_nodes_from(f"movie-{n}" for n in [0,1,2,3,4])
G.add_nodes_from(f"user-{n}" for n in [0,1,2])
```
Somehow you have to build unique nodes.  You need to map your existing labels to unique values. Those new unique values can be integers (my intent in the previous answer), but they can also be strings as this example shows.

**Reply by boothby** [reply]
My typical approach to this is to make tuples.  The simplest thing is `G.add_nodes_from(('user', i) for i in [0, 1, 2, 3, 4, 5, 6])`.  This avoids constructing and later parsing strings.

**Reply by Jeriousman** [reply]
Thank you for all your valuable help. I will try all!

**Reply by Jeriousman** [reply]
@dschult Thank you for your help. What I need exactly seems not really working in both cases of your examples. Your first approach is as this
```
G = nx.DiGraph()
G.add_nodes_from([0, 1, 2, 3, 4], type="movie")
G.add_nodes_from([0, 1, 2, 3, 4], type="user")
```
What I want is exactly the same node names (all starting from 0, 1, 2 ~ for movie and 0, 1, 2 ~ for user as well). But the codes above updates nodes [0, 1, 2, 3, 4] from movie type to user type then the movie type nodes are simply gone.

Your second approach keeps all the nodes, but naming 'movie-' and 'user-' are entailing node numbers which is not what I ultimately want. I guess what I ultimately want is not really how it is designed in NetworkX? Or is there another solution?

Thank you.

**Reply by dschult** [reply]
NetworkX is flexible in how nodes are constructed. I don't understand what it is that you want. But the solutions presented here -- especially @boothby 's solution -- seem like what you want. You talk a lot about what you don't want. But if you can figure out what you do want, instead of what you don't want, and then figure out how python could encode that information, then you can use it with NetworkX.

**Reply by Jeriousman** [reply]
Well, I cant be any more explicit about what I want as I even wrote a solid example I want up there. But obviously boothby offered one that would be what I want. I guess from there I should figure out how NetworkX exactly works in my goal.

**Reply by boothby** [reply]
I understand what you want, but unfortunately, that is impossible to do with NetworkX graphs.  I've frequently wanted, for example, bipartite graphs to support overlapping node labels between the sides.  But any time I've tried to think of how that would work in the generality we tend to provide, it goes to mush (for example, it would greatly complicate edge listing and that alone might force you into handling `(label, type)` tuples).  So, using tuples as labels is about as simple as it gets.
